DSharpPlus/Net/RestClient.cs

```diff
@@ -14,27 +15,35 @@
 
 namespace DSharpPlus.Net
 {
+    /// <summary>
+    /// Creates an instance of a Http client implementation.
+    /// </summary>
+    /// <param name="messageHandler">Message handler to use for the new Http client instance.</param>
+    /// <returns>Constructed Http client implementation.</returns>
+    public delegate HttpClient HttpClientFactoryDelegate(HttpMessageHandler messageHandler);
+
     /// <summary>
     /// Represents a client used to make REST requests.
     /// </summary>
     internal sealed class RestClient
     {
         private static UTF8Encoding UTF8 { get; } = new UTF8Encoding(false);
         private static Regex RouteArgumentRegex { get; } = new Regex(@":([a-z_]+)");
+        private static TimeSpan RateLimitResetBuffer { get; } = TimeSpan.FromMilliseconds(10);
 
         private BaseDiscordClient Discord { get; }
         private HttpClient HttpClient { get; }
-        private HashSet<RateLimitBucket> Buckets { get; }
-        private SemaphoreSlim RequestSemaphore { get; }
```

Comment:
Presumably this means you have implemented buckets to operate in an entirely atomic manner.